Add support for all CLP FFI functionality and refactor existing code:

davidlion opened the pull request:
- Add CLP encoding/decoding functionalities
- Add CLP IR wildcard query support
- Add basic GitHub templates and workflows
- Rewrite CLP IR serialization/deserialization functionalities to match the latest CLP glossary, and simplify resource management
- Add reader/writer interfaces for CLP IR
- Update formatting and lining tooling
- Switch to use static library

# Description
* all CLP encoding/decoding/searching FFI functionality
* simplify resource management and control flow
    * Go code will need to call Close() functions to avoid memory leaks, but code is much easier to understand and less error prone
* switch to static library
* rewrite/add documentation for all user-facing code
* rewrite terminology to match new CLP glossary
* update formatting + linting tooling
* github templates + workflows

# Validation performed
* Old/existing unit tests remade and passing.
* Added new unit tests.

Note, this is the same as #2 but from the beta branch to de-duplicate branches.

davidlion committed:
Large refactor and update

- Add all clp-ffi functions
- Add reader and writer
- Add unit tests
- Add formatting and linting
- Overhaul doc strings and terminology

davidlion committed:
Reintroduce __cplusplus ifdefs to headers so the correct names are in the header when used with a c++ compiler.

LinZhihao-723 commented on cpp/src/ffi_go/ir/LogTypes.hpp:
The name of this file is a bit confusing

LinZhihao-723 commented on cpp/src/ffi_go/ir/LogTypes.hpp:
```diff
@@ -0,0 +1,79 @@
+
+template <typename encoded_var_t>
+struct LogMessage {
+    auto reserve(size_t cap) -> void { m_logtype.reserve(cap); }
```
Shall we be more explicit that it reserves the storage for `m_logtype`?
```suggestion
    auto reserve_logtype(size_t cap) -> void { m_logtype.reserve(cap); }
```

davidlion replied:
This is done to mirror std:string

LinZhihao-723 commented on cpp/src/ffi_go/LogTypes.hpp:
The name of the file is confusing. Maybe name it `LogEventTypes`?

LinZhihao-723 commented on cpp/src/ffi_go/LogTypes.hpp:
```diff
@@ -0,0 +1,26 @@
+ * Mutating a field will invalidate the corresponding View (slice) stored in the
+ * ffi.LogEventView (without any warning or way to guard in Go).
+ */
+struct LogEvent {
```
It is a little strange that `LogEvent` only contains the message. For unstructured log events, it should contain a string message + a timestamp. But the timestamp is separate from the LogEvent class.

davidlion replied:
I see why this is confusing. Maybe we should add a comment and/or change the type name (`LogEventStorage`?).

There is no need to have any backing storage for a timestamp since it is passed by value to and from go. The field would essentially be dead code as we'd update it, but never read it.

LinZhihao-723 commented on cpp/src/ffi_go/ir/LogTypes.hpp:
```diff
@@ -0,0 +1,79 @@
+ * ir.Deserializer (without any warning or way to guard in Go).
+ */
+struct Deserializer {
+    ffi_go::LogEvent m_log_event;
```
as mentioned in `ffi_go::LogEvent` definition, it's strange that the timestamp is not a member of the log event.

davidlion replied:
See response in https://github.com/y-scope/clp-ffi-go/pull/3#discussion_r1564779651

LinZhihao-723 commented on cpp/src/ffi_go/ir/decoder.cpp:
```diff
@@ -0,0 +1,104 @@
+     * Generic helper for ir_decoder_decode_*_log_message
+     */
+    template <class encoded_var_view_t>
+    auto decode_log_message(
```
```suggestion
    [[nodiscard]] auto decode_log_message(
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/decoder.cpp:
```diff
@@ -0,0 +1,104 @@
+                    dict_var_end_offsets.m_data,
+                    dict_var_end_offsets.m_size
+            );
+        } catch (ffi::EncodingException& e) {
```
```suggestion
        } catch (ffi::EncodingException const& e) {
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/decoder.cpp:
```diff
@@ -0,0 +1,104 @@
+    }
+}  // namespace
+
+extern "C" auto ir_decoder_new() -> void* {
```
I'm not sure whether we should just use `extern "C" {}` to surround all these functions

davidlion replied:
This is a bit subjective... There is an argument for organization, but it is a little less error prone to be explicit about what is extern C. Also, mistakenly adding/removing something from an extern C block could have cryptic errors (relative to mistakenly adding/removing something from a namespace).

LinZhihao-723 replied:
Solved by using decoration macro

LinZhihao-723 commented on cpp/src/ffi_go/ir/decoder.h:
We may want to fix the indentation of this file. (changes indicated in the clang-format config file)

LinZhihao-723 commented on cpp/.clang-format:
```diff
@@ -1,26 +1,157 @@
+IndentAccessModifiers: false
+IndentCaseBlocks: false
+IndentCaseLabels: true
+IndentExternBlock: Indent
```
```suggestion
IndentExternBlock: false
```
This will disable indentations in the .h headers

davidlion replied:
We don't want to disable this globally.

LinZhihao-723 replied:
The reason is because CLion doesn't provide a good tracking of `extern C` blocks surrounded by `{ }`

LinZhihao-723 commented on cpp/src/ffi_go/ir/decoder.h:
```diff
@@ -0,0 +1,95 @@
+     * @param[in] vars Array of encoded variables
+     * @param[in] dict_vars String containing all dictionary variables concatenated
+     *     together
+     * @param[in] dict_var_end_offsets Array of offsets into dict_vars makring the
```
```suggestion
     * @param[in] dict_var_end_offsets Array of offsets into dict_vars marking the
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/decoder.h:
```diff
@@ -0,0 +1,95 @@
+     * @param[in] vars Array of encoded variables
+     * @param[in] dict_vars String containing all dictionary variables concatenated
+     *     together
+     * @param[in] dict_var_end_offsets Array of offsets into dict_vars makring the
```
```suggestion
     * @param[in] dict_var_end_offsets Array of offsets into dict_vars marking the
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/deserializer.h:
```diff
@@ -0,0 +1,165 @@
+    void ir_deserializer_close(void* ir_deserializer);
+
+    /**
+     * Given a CLP IR buffer (any encoding), attempt to deserialize a premable and
```
```suggestion
     * Given a CLP IR buffer (any encoding), attempt to deserialize a preamble and
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/deserializer.cpp:
```diff
@@ -0,0 +1,308 @@
+     * Generic helper for ir_deserializer_deserialize_*_log_event
+     */
+    template <class encoded_variable_t>
+    auto deserialize_log_event(
```
```suggestion
    [[nodiscard]] auto deserialize_log_event(
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/deserializer.cpp:
```diff
@@ -0,0 +1,308 @@
+     * Generic helper for ir_deserializer_deserialize_*_wildcard_match
+     */
+    template <class encoded_variable_t>
+    auto deserialize_wildcard_match(
```
```suggestion
    [[nodiscard]] auto deserialize_wildcard_match(
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/deserializer.h:
```diff
@@ -0,0 +1,165 @@
+     */
+    int ir_deserializer_deserialize_four_byte_wildcard_match(
+            ByteSpan ir_view,
+            void* ir_deserializer,
```
Same above

LinZhihao-723 commented on cpp/src/ffi_go/ir/deserializer.h:
```diff
@@ -0,0 +1,165 @@
+     */
+    int ir_deserializer_deserialize_eight_byte_log_event(
+            ByteSpan ir_view,
+            void* ir_deserializer,
```
Shall we explicitly ask the input type to be `Deserializer*` and leave the ptr cast to the caller?

davidlion replied:
This adds non-trivial complication as we would require interfacing between cgo and c++.

LinZhihao-723 commented on cpp/src/ffi_go/ir/deserializer.h:
```diff
@@ -0,0 +1,165 @@
+     */
+    int ir_deserializer_deserialize_four_byte_log_event(
+            ByteSpan ir_view,
+            void* ir_deserializer,
```
Same above

LinZhihao-723 commented on cpp/src/ffi_go/ir/deserializer.h:
```diff
@@ -0,0 +1,165 @@
+     */
+    int ir_deserializer_deserialize_eight_byte_wildcard_match(
+            ByteSpan ir_view,
+            void* ir_deserializer,
```
Same above

LinZhihao-723 commented on cpp/src/ffi_go/ir/deserializer.cpp:
```diff
@@ -0,0 +1,308 @@
+        size_t pos{0};
+        for (size_t i{0}; i < merged_query.m_end_offsets.m_size; i++) {
+            queries[i].first = query_view.substr(pos, end_offsets[i]);
+            queries[i].second = case_sensitivity[i];
```
```suggestion
            auto& [query_str_view, is_case_sensitive]{queries[i]};
            query_str_view = query_view.substr(pos, end_offsets[i]);
            is_case_sensitive = case_sensitivity[i];
```
How about refactoring like this for better readability?

LinZhihao-723 commented on cpp/src/ffi_go/ir/deserializer.cpp:
```diff
@@ -0,0 +1,308 @@
+            if (time_interval.m_lower > deserializer->m_timestamp) {
+                continue;
+            }
+            std::pair<bool, size_t> const match{query_fn(deserializer->m_log_event.m_log_message)};
+            if (match.first) {
+                size_t pos{0};
+                if (ErrorCode_Success != ir_buf.try_get_pos(pos)) {
+                    return static_cast<int>(IRErrorCode_Decode_Error);
+                }
+                *ir_pos = pos;
+                log_event->m_log_message.m_data = deserializer->m_log_event.m_log_message.data();
+                log_event->m_log_message.m_size = deserializer->m_log_event.m_log_message.size();
+                log_event->m_timestamp = deserializer->m_timestamp;
+                *matching_query = match.second;
+                return static_cast<int>(IRErrorCode_Success);
+            }
```
```suggestion
            auto const [has_matching_query, matching_query_idx]{
                    query_fn(deserializer->m_log_event.m_log_message)
            };
            if (false == has_matching_query) {
                continue;
            }
            size_t pos{0};
            if (ErrorCode_Success != ir_buf.try_get_pos(pos)) {
                return static_cast<int>(IRErrorCode_Decode_Error);
            }
            *ir_pos = pos;
            log_event->m_log_message.m_data = deserializer->m_log_event.m_log_message.data();
            log_event->m_log_message.m_size = deserializer->m_log_event.m_log_message.size();
            log_event->m_timestamp = deserializer->m_timestamp;
            *matching_query = matching_query_idx;
            return static_cast<int>(IRErrorCode_Success);
```

LinZhihao-723 reviewed (commented):
Add the first batch of comments on cpp decoder/deserializer implementations

LinZhihao-723 commented on cpp/src/ffi_go/ir/decoder.h:
```diff
@@ -0,0 +1,95 @@
+// NOLINTBEGIN(modernize-use-trailing-return-type)
+// NOLINTBEGIN(modernize-use-using)
+
+#ifdef __cplusplus
```
`extern "C"` block should not wrap header includes. Same for the rest of header files

LinZhihao-723 reviewed (changes requested):
Headers

LinZhihao-723 commented on .github/workflows/build.yml:
Questions:
1. why do we drop macos build?
2. can we run the test without setting `go_test_ir`? If not then the `go test ./...` in the end seems broken
3. we should have some `TODO` marked to update the workflow. It is not lining C++ code

davidlion replied:
1. Been too long, I'll test it again
2. Can you elaborate? The tests still run without setting `go_test_ir` but will skip the unit test requiring an existing ir file. Ideally, the unit tests are improved to not rely on existing IR files so we also don't need to store IR files in the repo.

davidlion replied:
For macos, someone needs to upload a pre-built library for CI to pass. Could you do it?

LinZhihao-723 replied:
kk, lgtm

LinZhihao-723 reviewed (commented):
Comments about the workflow

LinZhihao-723 commented on README.rst:
```diff
@@ -72,12 +76,12 @@ __ https://github.com/bazelbuild/rules_go/blob/master/docs/go/core/rules.md#go_l
-  compressed with zstd. This file's path is taken as the only argument to the
-  test and is supplied after ``-args``. It can be an absolute path or a path
-  relative to the ``ir`` directory.
+- Some of the ``ir`` package's tests currently requries an existing CLP IR file
```
```suggestion
- Some of the ``ir`` package's tests currently require an existing CLP IR file
```

LinZhihao-723 commented on README.rst:
Shall we move `Why not build with cgo?` section from `Testing` to `Building`?

LinZhihao-723 reviewed (changes requested):
Comments on the README

LinZhihao-723 commented on cpp/.clang-format:
```diff
@@ -1,26 +1,157 @@
+LambdaBodyIndentation: Signature
+LineEnding: LF
+MaxEmptyLinesToKeep: 1
 NamespaceIndentation: Inner
```
```suggestion
NamespaceIndentation: None
```
To sync with our latest style guide

LinZhihao-723 commented on cpp/src/ffi_go/ir/decoder.cpp:
```diff
@@ -0,0 +1,104 @@
+        typedef typename std::conditional<
+                std::is_same_v<Int64tSpan, encoded_var_view_t>,
+                ffi::eight_byte_encoded_variable_t,
+                ffi::four_byte_encoded_variable_t>::type encoded_var_t;
```
```suggestion
        using encoded_var_t = std::conditional<
                std::is_same_v<Int64tSpan, encoded_var_view_t>,
                ffi::eight_byte_encoded_variable_t,
                ffi::four_byte_encoded_variable_t>::type;
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/decoder.cpp:
```diff
@@ -0,0 +1,104 @@
+                ffi::eight_byte_encoded_variable_t,
+                ffi::four_byte_encoded_variable_t>::type encoded_var_t;
+        Decoder* decoder{static_cast<Decoder*>(ir_decoder)};
+        ffi_go::LogMessage& log_msg = decoder->m_log_message;
```
```suggestion
        auto& log_msg{decoder->m_log_message};
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/encoder.h:
```diff
@@ -0,0 +1,105 @@
+     * @param[out] vars Array of encoded variables
+     * @param[out] dict_vars String containing all dictionary variables concatenated
+     *     together
+     * @param[out] dict_var_end_offsets Array of offsets into dict_vars makring the
```
```suggestion
     * @param[out] dict_var_end_offsets Array of offsets into dict_vars marking the
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/encoder.h:
```diff
@@ -0,0 +1,105 @@
+     * @param[out] vars Array of encoded variables
+     * @param[out] dict_vars String containing all dictionary variables concatenated
+     *     together
+     * @param[out] dict_var_end_offsets Array of offsets into dict_vars makring the
```
```suggestion
     * @param[out] dict_var_end_offsets Array of offsets into dict_vars marking the
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/encoder.cpp:
```diff
@@ -0,0 +1,134 @@
+        typedef typename std::conditional<
+                std::is_same_v<Int64tSpan, encoded_var_view_t>,
+                ffi::eight_byte_encoded_variable_t,
+                ffi::four_byte_encoded_variable_t>::type encoded_var_t;
```
```suggestion
        using encoded_var_t std::conditional<
                std::is_same_v<Int64tSpan, encoded_var_view_t>,
                ffi::eight_byte_encoded_variable_t,
                ffi::four_byte_encoded_variable_t>::type;
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/encoder.cpp:
```diff
@@ -0,0 +1,134 @@
+                ffi::eight_byte_encoded_variable_t,
+                ffi::four_byte_encoded_variable_t>::type encoded_var_t;
+        Encoder<encoded_var_t>* encoder{static_cast<Encoder<encoded_var_t>*>(ir_encoder)};
+        LogMessage<encoded_var_t>& ir_log_msg = encoder->m_log_message;
```
```suggestion
        auto& ir_log_msg{encoder->m_log_message};
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/encoder.cpp:
```diff
@@ -0,0 +1,134 @@
+        ir_log_msg.reserve(log_message.m_size);
+
+        std::string_view const log_msg_view{log_message.m_data, log_message.m_size};
+        std::vector<int32_t> dict_var_offsets;
```
I think this vector can be a part of the ir_encoder storage. In that way, the vector can be reused to save some init/deinit and vector grow overhead.

davidlion replied:
If profiling shows this is an issue you could add it in a future PR. Otherwise, we should prefer local variables.

LinZhihao-723 replied:
sure, make senses to me

LinZhihao-723 commented on cpp/src/ffi_go/ir/encoder.cpp:
```diff
@@ -0,0 +1,134 @@
+
+        // dict_var_offsets contains begin_pos followed by end_pos of each
+        // dictionary variable in the message
+        int32_t prev_end_off = 0;
```
```suggestion
        int32_t prev_end_off{0};
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/encoder.cpp:
```diff
@@ -0,0 +1,134 @@
+        // dictionary variable in the message
+        int32_t prev_end_off = 0;
+        for (size_t i = 0; i < dict_var_offsets.size(); i += 2) {
+            int32_t const begin_pos = dict_var_offsets[i];
+            int32_t const end_pos = dict_var_offsets[i + 1];
```
```suggestion
            int32_t const begin_pos{dict_var_offsets[i]};
            int32_t const end_pos{dict_var_offsets[i + 1]};
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/encoder.cpp:
```diff
@@ -0,0 +1,134 @@
+            int32_t const begin_pos = dict_var_offsets[i];
+            int32_t const end_pos = dict_var_offsets[i + 1];
+            ir_log_msg.m_dict_vars.insert(
+                    ir_log_msg.m_dict_vars.begin() + prev_end_off,
+                    log_msg_view.begin() + begin_pos,
+                    log_msg_view.begin() + end_pos
```
```suggestion
                    ir_log_msg.m_dict_vars.cbegin() + prev_end_off,
                    log_msg_view.cbegin() + begin_pos,
                    log_msg_view.cbegin() + end_pos
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/serializer.h:
This file doesn't have `extern "C"` properly wrapped.

LinZhihao-723 commented on cpp/src/ffi_go/ir/serializer.h:
```diff
@@ -0,0 +1,109 @@
+void ir_serializer_close(void* ir_serializer);
+
+/**
+ * Given the fields of a CLP IR premable, serialize them into an IR byte stream
```
```suggestion
 * Given the fields of a CLP IR preamble, serialize them into an IR byte stream
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/serializer.h:
```diff
@@ -0,0 +1,109 @@
+ * @param[in] time_zone_id TZID timezone of the timestamps in the IR
+ * @param[out] ir_serializer_ptr Address of a new ir::Serializer
+ * @param[out] ir_view View of a IR buffer containing the serialized preamble
+ * @return ffi::ir_stream::IRErrorCode forwared from
```
```suggestion
 * @return ffi::ir_stream::IRErrorCode forwarded from
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/serializer.h:
```diff
@@ -0,0 +1,109 @@
+);
+
+/**
+ * Given the fields of a CLP IR premable, serialize them into an IR byte stream
```
```suggestion
 * Given the fields of a CLP IR preamble, serialize them into an IR byte stream
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/serializer.h:
```diff
@@ -0,0 +1,109 @@
+ * @param[in] time_zone_id TZID timezone of the timestamps in the IR
+ * @param[out] ir_serializer_ptr Address of a new ir::Serializer
+ * @param[out] ir_view View of a IR buffer containing the serialized preamble
+ * @return ffi::ir_stream::IRErrorCode forwared from
```
```suggestion
 * @return ffi::ir_stream::IRErrorCode forwarded from
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/serializer.h:
```diff
@@ -0,0 +1,109 @@
+ * @param[in] timestamp Timestamp of the log event to serialize
+ * @param[in] ir_serializer ir::Serializer object to be used as storage
+ * @param[out] ir_view View of a IR buffer containing the serialized log event
+ * @return ffi::ir_stream::IRErrorCode forwared from
```
```suggestion
 * @return ffi::ir_stream::IRErrorCode forwarded from
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/serializer.h:
```diff
@@ -0,0 +1,109 @@
+ *     IR stream
+ * @param[in] ir_serializer ir::Serializer object to be used as storage
+ * @param[out] ir_view View of a IR buffer containing the serialized log event
+ * @return ffi::ir_stream::IRErrorCode forwared from
```
```suggestion
 * @return ffi::ir_stream::IRErrorCode forwarded from
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/serializer.cpp:
```diff
@@ -0,0 +1,153 @@
+            epoch_time_ms_t timestamp_or_delta,
+            void* ir_serializer,
+            ByteSpan* ir_view
+    ) -> int {
```
Seems like we don't check whether the input ptrs are null. Shall we add the check?
If yes, we should also add the check wherever a passed-in ptr gets dereferenced.

LinZhihao-723 commented on cpp/src/ffi_go/ir/serializer.cpp:
```diff
@@ -0,0 +1,153 @@
+        StringView time_zone_id,
+        void** ir_serializer_ptr,
+        ByteSpan* ir_view
+) -> int {
```
missing nullptr check for `ir_serializer_ptr` and `ir_view` (which exists in the four byte version)

LinZhihao-723 commented on cpp/src/ffi_go/ir/serializer.cpp:
```diff
@@ -0,0 +1,153 @@
+    return static_cast<int>(IRErrorCode_Success);
+}
+
+extern "C" auto ir_serializer_serialize_four_byte_preamble(
```
The function has a lot of duplications with `ir_serializer_serialize_eight_byte_preamble`.
Shall we make a template class instead?

davidlion replied:
You can refactor this code if you'd like (ensure it works with cgo), but I don't have time currently. If you're changing it in your PR then maybe refactoring this is unnecessary.

davidlion replied:
Nvm done.

LinZhihao-723 commented on cpp/src/ffi_go/ir/serializer.h:
```diff
@@ -0,0 +1,109 @@
+ * @return ffi::ir_stream::IRErrorCode forwared from
+ *     ffi::ir_stream::four_byte_encoding::encode_preamble
+ */
+int ir_serializer_serialize_four_byte_preamble(
```
I think the naming of these two functions (`ir_serializer_serialize_four_byte_preamble` and `ir_serializer_serialize_eight_byte_preamble`) is unclear. Essentially, these functions not just serialize the preamble bytes, but creating a new serializer by allocating heap memory + initialization. We should be more explicit to call it with sth like `ir_serializer_create_xxxx_byte_serializer_with_preamble`

LinZhihao-723 reviewed (changes requested):
Add comments on cpp implementations of encoders and serializers

LinZhihao-723 commented on cpp/src/ffi_go/search/wildcard_query.h:
```diff
@@ -0,0 +1,78 @@
+     * @param[in] ptr Address of a new std::string
+     * @return New string holding cleaned query
+     */
+    StringView wildcard_query_clean(StringView query, void** ptr);
```
Some comments:
1. we should move this function ahead of the deleter. It's confusing to first read `delete`
2. I'm not sure if this is a good function. It is not simply cleaning the given string but also creating underlying storage. Maybe we should name it `wildcard_query_get_cleaned_query_str`, and also name `wildcard_query_delete` to `wildcard_query_delete_cleaned_query_str`
3. The comments referring to `CLP src/string_utils.hpp` will be outdated if we update CLP core dependency. Maybe we should explicitly explain why we need this function in the doc string instead of referring to CLP core.

LinZhihao-723 commented on cpp/src/ffi_go/search/wildcard_query.cpp:
```diff
@@ -0,0 +1,30 @@
+
+extern "C" auto wildcard_query_match(StringView target, WildcardQueryView query) -> int {
+    return static_cast<int>(wildcard_match_unsafe(
+            std::string_view{target.m_data, target.m_size},
+            std::string_view{query.m_query.m_data, query.m_query.m_size},
+            static_cast<bool>(query.m_case_sensitive)
```
```suggestion
            {target.m_data, target.m_size},
            {query.m_query.m_data, query.m_query.m_size},
            query.m_case_sensitive
```
Do we need `static_cast<bool>`? Is it added because the original type was declared with `<stdbool.h>`?

LinZhihao-723 commented on cpp/src/ffi_go/search/wildcard_query.cpp:
```diff
@@ -0,0 +1,30 @@
+}
+
+extern "C" auto wildcard_query_clean(StringView query, void** ptr) -> StringView {
+    auto* clean = new std::string{
+            clean_up_wildcard_search_string(std::string_view{query.m_data, query.m_size})};
```
```suggestion
    auto* clean{new std::string{
            clean_up_wildcard_search_string(std::string_view{query.m_data, query.m_size})
    }};
```

LinZhihao-723 reviewed (changes requested):
Comments on the wildcard query C++ implementations.

LinZhihao-723 commented on ffi/ffi.go:
```diff
@@ -1,69 +1,33 @@
+// The ffi package contains the general types representing log events  created
+// and used by logging functions or libraries. In other words log events with no
+// sort of CLP encoding or serializing.
```
```suggestion
// Package ffi defines types for log events used in logging functions and
// libraries, without CLP encoding or serialization.
```

LinZhihao-723 commented on ffi/ffi.go:
```diff
@@ -1,69 +1,33 @@
+// A LogMessageView is a LogMessage that is backed by C++ allocated memory
+// rather than the Go heap. A LogMessageView, x, is valid when returned and will
+// remain valid until a new LogMessageView is returned by the same object (e.g.
+// an ir.Deserializer) that retuend x.
```
```suggestion
// an ir.Deserializer) that returns x.
```

LinZhihao-723 commented on ffi/ffi.go:
```diff
@@ -1,69 +1,33 @@
 	Timestamp EpochTimeMs
 }
+
+// The underlying memory of LogEventView is C-allocated and owned by the object
+// (e.g. reader, desializer, etc) that returned it. Using an existing
+// LogEventView after a new view has been returned by the same producing object
+// is undefined (different producing objects own their own memory for views).
```
```suggestion
// LogEventView memory is allocated and owned by the C++ object (e.g., reader,
// deserializer) that returns it. Reusing a LogEventView after the same object
// has issued a new view leads to undefined behavior, as different objects
// manage their own memory independently.
```

LinZhihao-723 reviewed (changes requested):
Comments on `ffi/ffi.go`. All about the doc strings.

davidlion committed:
Apply suggestions from code review

- fix typos
- update NamespaceIndentation to None
- add [[nodiscard]]
- typedef -> using
- add const
- use fancy auto syntax

Co-authored-by: Lin Zhihao <59785146+LinZhihao-723@users.noreply.github.com>

davidlion committed:
Apply suggestions from code review

- drop static cast

Co-authored-by: Lin Zhihao <59785146+LinZhihao-723@users.noreply.github.com>

davidlion committed:
Improve writing

davidlion committed:
Fix extern C header ifdef to not wrap include statements.

davidlion committed:
Add macos back in workflow

davidlion committed:
LogTypes.hpp -> types.hpp

davidlion committed:
Update cpp/src/ffi_go/ir/decoder.cpp

Co-authored-by: Lin Zhihao <59785146+LinZhihao-723@users.noreply.github.com>

LinZhihao-723 commented on cpp/src/ffi_go/ir/encoder.cpp:
```diff
@@ -0,0 +1,134 @@
+            StringView* dict_vars,
+            Int32tSpan* dict_var_end_offsets
+    ) -> int {
+        using encoded_var_t std::conditional<
```
```suggestion
        using encoded_var_t = std::conditional<
```
Sorry for missing `=` in the previous comment

LinZhihao-723 committed:
Update cpp/src/ffi_go/ir/encoder.cpp

LinZhihao-723 commented on cpp/src/ffi_go/ir/decoder.cpp:
```diff
@@ -0,0 +1,104 @@
+            void* ir_decoder,
+            StringView* log_msg_view
+    ) -> int {
+        using encoded_var_t = std::conditional<
```
```suggestion
        using encoded_var_t = typename std::conditional<
```

LinZhihao-723 commented on cpp/src/ffi_go/ir/encoder.cpp:
```diff
@@ -0,0 +1,134 @@
+            StringView* dict_vars,
+            Int32tSpan* dict_var_end_offsets
+    ) -> int {
+        using encoded_var_t = std::conditional<
```
```suggestion
        using encoded_var_t = typename std::conditional<
```

LinZhihao-723 reviewed (changes requested):
According to the cpp standard, for dependent names we need `typename` disambiguator to explicitly tell the compiler this is a type name: https://en.cppreference.com/w/cpp/language/dependent_name
Even though the code still gets compiled in some compilers without `typename`, it is a good practice to add it anyways.

LinZhihao-723 committed:
Adding `typename` for dependent names for disambiguration purpose.

davidlion committed:
Refactor c++ and add missing null checks.

davidlion committed:
Rename LogEvent to LogEventStorage.

LinZhihao-723 commented on ir/serializer.go:
```diff
@@ -0,0 +1,172 @@
+// cptr holds a reference to the underlying C++ objected used as backing storage
+// for the Views returned by the serializer. Close must be called to free this
+// underlying memory and failure to do so will result in a memory leak.
+type commonSerializer struct {
+	tsInfo TimestampInfo
+	cptr   unsafe.Pointer
+}
```
I understand that these structs are private to the package. However, shall we declare them before we define the `FourByteSerializer` and `EightByteSerialzier` creators?

davidlion replied:
Lets maintain the public -> private lexical ordering for now. This can be improved with a refactor later.

LinZhihao-723 commented on ir/serializer.go:
```diff
@@ -0,0 +1,172 @@
+}
+
+// Close will delete the underlying C++ allocated memory used by the
+// deserializer. Failure to call Close will result in a memory leak.
```
```suggestion
// Closes the serializer by releasing the underlying C++ allocated memory.
// Failure to call Close will result in a memory leak.
```

LinZhihao-723 commented on ir/serializer.go:
```diff
@@ -0,0 +1,172 @@
+	commonSerializer
+}
+
+// SerializeLogEvent attempts to serialize the log event, event, into a eight
```
```suggestion
// SerializeLogEvent attempts to serialize the log event, event, into an eight
```

LinZhihao-723 commented on ir/writer.go:
```diff
@@ -0,0 +1,135 @@
+//   - success: number of bytes written, nil
+//   - error: number of bytes written (can be 0), error propagated from
+//     [SerializeLogEvent] or [bytes.Buffer.Write]
+func (self *Writer) Write(event ffi.LogEvent) (int, error) {
```
How do we handle the case when Serialization succeeded but `bytes.Buffer.Write` failed?
We don't provide users with an API for a retry, but calling `Write` again may lead to incorrect behavior (for example, the reference timestamp has already changed by the previous serialization call)

davidlion replied:
Added an explanation in a comment to mention that `bytes.Buffer.Write` will always return `nil` for `err` (https://pkg.go.dev/bytes#Buffer.Write) and possible solutions if this ever changes.

LinZhihao-723 replied:
lgtm

LinZhihao-723 commented on ir/writer.go:
```diff
@@ -0,0 +1,135 @@
+// [self.Bytes] and [self.Reset] to manually handle the buffer's contents before
+// continuing. Returns:
+//   - success: number of bytes written, nil
+//   - error: number of bytes written, error propagated from [bytes.Buffer.WriteTo]
```
```suggestion
//   - error: number of bytes written, error propagated from
//     [bytes.Buffer.WriteTo]
```

LinZhihao-723 reviewed (changes requested):
Comments on `ir/writer.go` and `ir/serializer.go`

LinZhihao-723 commented on ir/ir.go:
```diff
@@ -1,63 +1,53 @@
+// analyzed using APIs provided in this package.
 //
 // [CLP]: https://github.com/y-scope/clp
 // [Uber blog]: https://www.uber.com/blog/reducing-logging-cost-by-two-orders-of-magnitude-using-clp/
```
Technically this line exceeds 100 chars. If wrapping it into a new line doesn't break the doc site we should do it

LinZhihao-723 commented on ir/ir.go:
```diff
@@ -1,63 +1,53 @@
+// ir.BufView is a slice of CLP IR backed by C++ allocated memory rather than
+// the Go heap. A BufView, x, is valid when returned and will remain valid until
+// a new BufView is returned by the same object (e.g. an [ir.Serializer]) that
+// retuend x.
```
```suggestion
// ir.BufView represents a slice of CLP IR, utilizing memory allocated by C++
// instead of the Go heap. A BufView, denoted as x, is valid upon being returned
// and maintains its validity until the same object (e.g., an [ir.Serializer])
// that issued x returns a new BufView.
```

LinZhihao-723 commented on ir/ir.go:
```diff
@@ -1,63 +1,53 @@
+// A ir.LogMessageView is a [ir.LogMessage] that is backed by C++ allocated
+// memory rather than the Go heap. A LogMessageView, x, is valid when returned
+// and will remain valid until a new LogMessageView is returned by the same
+// object (e.g.  an [ir.Encoder]) that retuend x.
```
```suggestion
// ir.LogMessageView is a [ir.LogMessage] using memory allocated by C++ instead
// of the Go heap. A LogMessageView, denoted as x, is valid upon being returned
// and maintains its validity until the same object (e.g., an [ir.Encoder])
// that issued x returns a new LogMessageView.
```

LinZhihao-723 commented on ir/deserializer.go:
```diff
@@ -0,0 +1,284 @@
+	}
+
+	var tsInfo TimestampInfo
+	if tsPat, ok := metadata["TIMESTAMP_PATTERN"].(string); ok {
+		tsInfo.Pattern = tsPat
+	}
+	if tsSyn, ok := metadata["TIMESTAMP_PATTERN_SYNTAX"].(string); ok {
+		tsInfo.PatternSyntax = tsSyn
+	}
+	if tzid, ok := metadata["TZ_ID"].(string); ok {
+		tsInfo.TimeZoneId = tzid
+	}
+
+	var deserializer Deserializer
+	if 1 == irEncoding {
+		var refTs ffi.EpochTimeMs = 0
+		if tsStr, ok := metadata["REFERENCE_TIMESTAMP"].(string); ok {
```
1. shall we make some constants/variables for the metadata field names?
2. we are not checking whether the IR stream version is supported. This is what we do in Python ffi: https://github.com/y-scope/clp-ffi-py/blob/main/src/clp_ffi_py/ir/native/decoding_methods.cpp#L204. The difference is that we decoded everything in the C++ layer. Maybe we should add a TODO here to ensure version numbers are properly validated.

davidlion replied:
1. will do
2. How about we return a version error from c++ in `ir_deserializer_new_deserializer_with_preamble`?

davidlion replied:
For 2 we will leave a TODO as sync'ing to the new CLP version will change this (on top of requiring other changes).

LinZhihao-723 replied:
sgtm

LinZhihao-723 commented on ir/deserializer.go:
```diff
@@ -0,0 +1,284 @@
+)
+
+// A Deserializer exports functions to deserialize log events from a CLP IR byte
+// stream. Deserializatoin functions take an IR buffer as input, but how that
```
```suggestion
// stream. Deserialization functions take an IR buffer as input, but how that
```

LinZhihao-723 commented on ir/deserializer.go:
```diff
@@ -0,0 +1,284 @@
+	return deserializeLogEvent(self, irBuf)
+}
+
+func (self *eightByteDeserializer) DeserializeWildcardMatch(
```
maybe we should move this function following `(self *eightByteDeserializer) DeserializieLogEvent`

LinZhihao-723 commented on ir/deserializer.go:
```diff
@@ -0,0 +1,284 @@
+) (*ffi.LogEventView, int, int, error) {
+	return deserializeWildcardMatch(self, irBuf, timeInterval, mergedQuery)
+}
+
```
```suggestion

// DeserializeWildcardMatch attempts to read the next log event from the IR
// stream in irBuf that matches mergedQuery within timeInterval. It returns the
// deserialized [ffi.LogEventView], the position read to in irBuf (the end of
// the log event in irBuf), the index of the matched query in mergedQuery,
// and an error. On error returns:
//   - nil *ffi.LogEventView
//   - 0 position
//   - -1 index
//   - [IrError] error: CLP failed to successfully deserialize
//   - [EndOfIr] error: CLP found the IR stream EOF tag
```

LinZhihao-723 commented on ir/deserializer.go:
```diff
@@ -0,0 +1,284 @@
+) (*ffi.LogEventView, int, error) {
+	return deserializeLogEvent(self, irBuf)
+}
+
```
```suggestion

// DeserializeWildcardMatch attempts to read the next log event from the IR
// stream in irBuf that matches mergedQuery within timeInterval. It returns the
// deserialized [ffi.LogEventView], the position read to in irBuf (the end of
// the log event in irBuf), the index of the matched query in mergedQuery,
// and an error. On error returns:
//   - nil *ffi.LogEventView
//   - 0 position
//   - -1 index
//   - [IrError] error: CLP failed to successfully deserialize
//   - [EndOfIr] error: CLP found the IR stream EOF tag
```

LinZhihao-723 commented on search/wildcard_query.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+// A timestamp interval of [m_lower, m_upper).
+type TimestampInterval struct {
```
Do we provide a way to create a default timestamp interval that covers the entire supported timestamp range?

davidlion replied:
This seems unnecessary as constructing the equivalent struct would also only be one line of code.

LinZhihao-723 replied:
But do we have TS_MIN and TS_MAX defined?

davidlion replied:
Why would these be different from int min/max?

LinZhihao-723 replied:
we defined the timestamp type to `EpochTimeMs` instead of explicitly using int. It might not be obvious for users what underlying type it is. It also introduces problems if we change the underlying type in the future.

LinZhihao-723 commented on search/wildcard_query.go:
```diff
@@ -0,0 +1,89 @@
+// query is case sensitive or not. The fields must be accessed through getters
+// to ensure that the query string remains clean/safe after creation by
+// NewWildcardQuery.
+// (Copied from clp/components/core/src/string_utils.hpp)
```
```suggestion
```
we could remove it in case it's relative path is out-dated

davidlion replied:
As long as the reader looks at the clp version used in this commit it can't be out-dated. It is also important for future readers to know the origin in case there were meaningful changes applied to the source.

LinZhihao-723 replied:
My concern is this comment can be outdated if we update CLP dependency version. When you update the underlying go ffi implementation, it is easy to miss we have a comment on the Go side pointing to C++ code (could happen where both the author and the reviewer are not aware of this comment).

davidlion replied:
It is a good point that the advantage of keeping the comment likely doesn't outweigh the potential confusion.

LinZhihao-723 reviewed (changes requested):
Comments on:
- ir/ir.go
- ir/decoder.go
- ir/deserializer.go
- search/wildcard_query.go

LinZhihao-723 commented on ir/reader.go:
```diff
@@ -122,26 +132,72 @@ func (self *IrReader) ReadToFunc(r io.Reader, f func(ffi.LogEvent) bool) (ffi.Lo
+// the buffer, it is grown by 1.5x if more than 1/4th of it is unconsumed IR.
+// Forwards the return of [ioReader.Read].
+func (self *Reader) fillBuf() (int, error) {
+	if (self.end - self.start) > len(self.buf)>>2 {
```
we could make a variable to store `len(self.buf)` for better readablity since it is referenced multiple times

davidlion replied:
It is debatable whether `buf_len` is more readable than `len(self.buf)` and it is less error prone to use `len` directly rather than a local variable that could become stale if `buf` is mutated.

LinZhihao-723 replied:
sure

LinZhihao-723 commented on ir/reader.go:
```diff
@@ -122,26 +132,72 @@ func (self *IrReader) ReadToFunc(r io.Reader, f func(ffi.LogEvent) bool) (ffi.Lo
-// for [ffi.LogEvent.Msg] and prefix. Errors are propagated from ReadNextLogEvent.
-func (self *IrReader) ReadToPrefix(r io.Reader, prefix []byte) (ffi.LogEvent, error) {
-	return self.ReadToFunc(r, func(e ffi.LogEvent) bool { return bytes.HasPrefix(e.Msg, prefix) })
+// fillBuf shifts the remaining valid IR in [Reader.buf] to the front and then
+// calls [ioReader.Read] to fill the remainder with more IR. Before reading into
+// the buffer, it is grown by 1.5x if more than 1/4th of it is unconsumed IR.
+// Forwards the return of [ioReader.Read].
```
This heuristic seems ok. But in Python libraries (both ffi and loglib), the heuristic we used to grow the buffer is "double the buffer capacity if more than half of the bytes are unconsumed". Any reasons why we choose a different heuristic here? And do we want to be consistent with the Python libraries?

LinZhihao-723 commented on ir/reader.go:
```diff
@@ -122,26 +132,72 @@ func (self *IrReader) ReadToFunc(r io.Reader, f func(ffi.LogEvent) bool) (ffi.Lo
-// time. Errors are propagated from ReadNextLogEvent.
-func (self *IrReader) ReadToEpochTime(r io.Reader, time ffi.EpochTimeMs) (ffi.LogEvent, error) {
-	return self.ReadToFunc(r, func(e ffi.LogEvent) bool { return e.Timestamp >= time })
+// Read the CLP IR stream until a [ffi.LogEventView] is greater than or equal to
+// the given timestamp. Errors are propagated from [ReadToFunc].
+func (self *Reader) ReadToEpochTime(
+	time ffi.EpochTimeMs,
+) (*ffi.LogEventView, error) {
+	return self.ReadToFunc(func(event *ffi.LogEventView) bool { return event.Timestamp >= time })
+}
+
+// Read the CLP IR stream until [strings/Contains] returns true for a
+// [ffi.LogEventView] and the given sub string. Errors are propagated from
+// [ReadToFunc].
+func (self *Reader) ReadToContains(substr string) (*ffi.LogEventView, error) {
+	fn := func(event *ffi.LogEventView) bool {
+		return strings.Contains(event.LogMessageView, substr)
+	}
+	return self.ReadToFunc(fn)
+}
+
+// Read the CLP IR stream until [strings/HasPrefix] returns true for a
+// [ffi.LogEventView] and the given prefix. Errors are propagated from
+// [ReadToFunc].
+func (self *Reader) ReadToPrefix(prefix string) (*ffi.LogEventView, error) {
+	fn := func(event *ffi.LogEventView) bool {
+		return strings.HasPrefix(event.LogMessageView, prefix)
+	}
+	return self.ReadToFunc(fn)
 }
```
The implementations of these functions are functionally correct. But they can be more efficient if we create a wildcard query and execute a query search.

davidlion replied:
This can be another PR.

LinZhihao-723 replied:
According to our discussion, the string methods can outperform wildcard searches since it could have better time complexity. We could rewrite these methods in the future PRs if needed.

LinZhihao-723 commented on ir/reader.go:
```diff
@@ -122,26 +132,72 @@ func (self *IrReader) ReadToFunc(r io.Reader, f func(ffi.LogEvent) bool) (ffi.Lo
+// CLP IR.
+func (self *Reader) read() (int, error) {
+	n, err := self.ioReader.Read(self.buf[self.end:])
+	self.end += n
+	if nil != err && io.EOF != err {
+		return n, err
+	}
```
```suggestion
	if nil != err && io.EOF != err {
		return n, err
	}
	self.end += n
```
Increment `end` after error handling

davidlion replied:
The current code defends against the case where `n != 0 and nil != err`.

LinZhihao-723 reviewed (changes requested):
Comments on ir/reader.go

LinZhihao-723 committed:
Apply suggestions from code review

davidlion committed:
Apply suggestions from code review

Improves/rewrites/fixes many doc strings

Co-authored-by: Lin Zhihao <59785146+LinZhihao-723@users.noreply.github.com>

LinZhihao-723 committed:
Refactor...

LinZhihao-723 committed:
Using api decoration.

LinZhihao-723 committed:
Apply clang-tidy suggestions on src/ffi_go/ir/*

LinZhihao-723 committed:
Apply clang-tidy suggestions on regular files

LinZhihao-723 committed:
Update github workflow to specify Go versions.

davidlion committed:
Workflow debugging.

davidlion committed:
Add api_decoration header.

davidlion committed:
Fix linting.

davidlion committed:
Fix workflows hopefully.

davidlion committed:
Removed confusing comment.

LinZhihao-723 committed:
Add darwin arm64 lib

davidlion committed:
Added helper function to reader to perform a wildcard query with the max time interval.

LinZhihao-723 reviewed (approved):
Need a multi-line commit message like this PR: https://github.com/y-scope/clp/pull/301

davidlion commented:
how about:

Add all CLP FFI functionality and refactor existing code: (#3)
- Add CLP IR encoding/decoding 
- Add CLP IR wildcard query
- Add basic GitHub templates and workflows
- Rewrite CLP IR serialization/deserialization functionalities to match the latest CLP glossary, and simplify resource management
- Add reader/writer interfaces for CLP IR
- Provide pre-built static libraries
